zerver/lib/markdown/__init__.py

```diff
@@ -601,6 +602,32 @@ def run(self, root: Element) -> None:
             img.set("src", get_camo_url(url))
 
 
+class InlineVideoProcessor(markdown.treeprocessors.Treeprocessor):
+    """
+    Rewrite inline video tags to serve external content via Camo.
+
+    This rewrites all video, except ones that are served from the current
+    realm or global STATIC_URL. This is to ensure that each realm only loads
+    videos that are hosted on that realm or by the global installation,
+    avoiding information leakage to external domains or between realms. We need
+    to disable proxying of videos hosted on the same realm, because otherwise
+    we will break videos in /user_uploads/, which require authorization to
+    view.
+    """
+
+    def run(self, root: Element) -> None:
+        # Get all URLs from the blob
+        found_videos = walk_tree(root, lambda e: e if e.tag == "video" else None)
+        for video in found_videos:
+            url = video.get("src")
+            assert url is not None
+            if is_static_or_current_realm_url(url, self.md.zulip_realm):
+                # Don't rewrite videos on our own site (e.g. user uploads).
+                continue
+            video.set("src", get_camo_url(url))
```

Comment:
Can we add a comment here documenting what properties we expect the camo version of the video to have? Do we have any guarantees on its size? 

One thing I'm worried about is someone linking to a 500MB video and that making a Zulip stream unvisitable, and I'd like to have a clearly written description in the code of what mechanism will prevent that.

Reply:
What's the difference between the `src` and the `data-original-src` -- it sounds like both will be full-size copies of the video, just loaded from different origins?

Reply:
The idea is to let camo load the preview, but the video plays from the original source, so our servers don't have to worry about traffic.

Reply:
I see, OK, that makes sense. Definitely deserves a comment.

Reply:
It looks like this now has a comment on it, so I'll resolve this thread.